lib/parameter_popup.dart

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+
+library dartpad.parameter_popup;
+
+import 'dart:convert';
+import 'dart:html';
+import 'dart:math' as math;
+
+import 'package:dart_pad/dartservices_client/v1.dart';
+import 'package:dart_pad/editing/editor.dart';
+import 'package:dart_pad/context.dart';
+import 'package:dart_pad/services/common.dart';
+
+
+class ParameterPopup {
+  DartservicesApi servicesApi;
+  Context context;
+  Editor editor;
+  static const List parKeys = const[
+    KeyCode.COMMA, KeyCode.NINE, KeyCode.ZERO,
+    KeyCode.LEFT, KeyCode.RIGHT, KeyCode.UP, KeyCode.DOWN];
+
+  HtmlEscape sanitizer = const HtmlEscape();
+
+  ParameterPopup(this.servicesApi, this.context, this.editor) {
+    document.onKeyDown.listen((e) => handleKeyDown(e));
+    document.onKeyUp.listen((e) => handleKeyUp(e));
+    document.onClick.listen((e) => handleClick());
+  }
+
+  bool get parPopupActive => querySelector(".parameter-hints") != null;
+
+  void remove() {
+    document.body.children.remove(querySelector(".parameter-hints"));
+  }
+
+  void handleKeyDown(KeyboardEvent e) {
+    if (e.keyCode == KeyCode.ENTER) {
+      _lookupParameterInfo();
+    }
+  }
+
+  void handleKeyUp(KeyboardEvent e) {
+    if (parPopupActive || parKeys.contains(e.keyCode)) {
+      _lookupParameterInfo();
+    }
+  }
+
+  void handleClick() {
+    if (context.focusedEditor != 'dart' || !editor.hasFocus) {
+      remove();
+      return;
+    }
+    _lookupParameterInfo();
+  }
+
+  void _lookupParameterInfo() {
+    int offset = editor.document.indexFromPos(editor.document.cursor);
+    String source = editor.document.value;
+    Map<String, int> parInfo = _parameterInfo(source, offset);
+
+    if (parInfo == null) {
+      remove();
+      return;
+    }
+
+    int openingParenIndex = parInfo["openingParenIndex"], parameterIndex = parInfo["parameterIndex"];
+    offset = openingParenIndex - 1;
+
+    // We request documentation info of what is before the parenthesis.
+    SourceRequest input = new SourceRequest()
+      ..source = source
+      ..offset = offset;
+
+    servicesApi.document(input).timeout(serviceCallTimeout)
+        .then((DocumentResponse result) {
+          if (!result.info.containsKey("parameters")) {
+            remove();
+            return;
+          }
+          List parameterInfo = result.info["parameters"] as List;
+          String outputString = "";
+          if (parameterInfo.length == 0) {
```

Comment:
You can; it might be slightly more performant. I wouldn't make the decision based on performance though; whatever you think reads better.